### PR TITLE
FOSSA scan enabled

### PR DESCRIPTION
**Pull request by idvoretskyi**
ref.: https://github.com/cncf/foundation/issues/109

Signed-off-by: Ihor Dvoretskyi <ihor@linux.com>

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Thanks!

I'm getting the following when trying a dry-run:

```
FATAL Could not analyze: ERROR: none of [npm-shrinkwrap.json package-lock.json] found at root of node project
```

Seems like FOSSA is set up to use NPM by default, is it possible to make it read the yarn state instead?

**Comment by idvoretskyi** [comment]
@Rugvip let's try again, I've updated the config file.

Also, FOSSA should use yarn by default as well - https://docs.fossa.com/docs/javascript

**Comment by idvoretskyi** [comment]
@Rugvip also, it seems I had some formatting issues in the YAML file, should be fixed now.

**Comment by Fox32** [comment]
We are using fossa for backstage too, maybe this helps you:

```yaml
on:
  push:
    branches:
      - master
  pull_request:

jobs:
  fossa-license-analyze:
    runs-on: ubuntu-latest
    timeout-minutes: 60
    steps:
      - name: Checkout
        uses: actions/checkout@v2.3.4

      - name: Setup node
        uses: actions/setup-node@v2.1.4
        with:
          node-version: '14'

      - name: Install yq
        run: sudo snap install yq

      - name: Install Fossa
        run: "curl -H 'Cache-Control: no-cache' https://raw.githubusercontent.com/fossas/fossa-cli/master/install.sh | bash"

      - name: Configure Fossa
        # The --option flag for fossa init does not work yet, see https://github.com/fossas/fossa-cli/issues/614
        run:  "fossa init --project PROJECT_NAME && yq e '.analyze.modules[].options.strategy = \"yarn-list\"' -i .fossa.yml"

      - name: Show config
        run: cat .fossa.yml

      - name: Fossa Analyze
        env:
          FOSSA_API_KEY: ${{ secrets.FOSSA_API_KEY }}
        run: fossa analyze --branch "$GITHUB_REF"

      # In case you are interested in the scan result:
      #- name: Fossa Test
      # env:
      #   FOSSA_API_KEY: ${{ secrets.FOSSA_API_KEY }}
      # run: fossa test --project "PROJECT_NAME" --branch "$GITHUB_REF" --debug
```

File: .github/workflows/fossa.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  pull_request:
+    branches: [master]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x, 14.x, 15.x]
```
Comment (Rugvip) [inline .github/workflows/fossa.yml:14]: Should be enough to just run towards 14 right?

File: .github/workflows/fossa.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+name: FOSSA
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
```
Comment (Rugvip) [inline .github/workflows/fossa.yml:6]: Is the purpose of running this towards PRs to validate the dependencies that they add?

File: .github/workflows/fossa.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  - uses: actions/checkout@v2
+  - uses: borales/actions-yarn@v2.0.0
+    with:
+      cmd: install
+  - uses: borales/actions-yarn@v2.0.0
+    with:
+      cmd: build
+  - uses: borales/actions-yarn@v2.0.0
+    with:
+      cmd: test
```
Comment (Rugvip) [inline .github/workflows/fossa.yml:26]: install, build and test are run in a separate workflow, so let's not duplicate those here. Looking at the suggestion from @Fox32 it seems like they're not needed anyway

**Review by Rugvip — COMMENTED** [review]
Thanks! Gonna close and reopen to have the build triggered, wanna make sure it goes through

File: .github/workflows/fossa.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
+
+steps:
```
Comment (Rugvip) [inline .github/workflows/fossa.yml:16]: Oh this entire block should be indented once. Prolly why it's not triggering correctly

**Comment by dhenneke** [comment]
```
============================================================

    View FOSSA Report:
    https://app.fossa.com/projects/custom+162%2FPROJECT_NAME/refs/branch/refs%2Fpull%2F4077%2Fmerge/618a4e54bd602782a3b549c67afae576a229c6a8

============================================================
```

I doubt that `PROJECT_NAME` is the correct name here 😉

**Comment by Rugvip** [comment]
@dhenneke shite, though I removed that :grin:

**Comment by idvoretskyi** [comment]
@Rugvip oh, awesome, thanks!

**Comment by Rugvip** [comment]
@idvoretskyi Wondering, where can we verify the result?

**Comment by idvoretskyi** [comment]
See here - https://github.com/backstage/backstage/runs/1736600751?check_suite_focus=true#step:7:67

**Comment by Rugvip** [comment]
@idvoretskyi Gives me a 404 😅

**Comment by freben** [comment]
Me too, but I needed to sign in and now I'm waiting to be added to the Spotify org there ... were you signed in when you got the 404?

**Comment by freben** [comment]
@idvoretskyi We keep getting 502 errors blocking builds.

```
INFO Uploading analysis...
FATAL Error during upload: Error response from API: <html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

https://github.com/backstage/backstage/pull/4177/checks?check_run_id=1740390959

**Comment by benjdlambert** [comment]
I managed to get one working by rebuilding this morning, but it seems to break more often than not with all our packages.

**Comment by idvoretskyi** [comment]
@freben that's weird, seems like a FOSSA issue.

Let me check this with the FOSSA team.

**Comment by idvoretskyi** [comment]
@freben it seems to work here, I'm a bit confused - https://github.com/backstage/backstage/runs/1741859298?check_suite_focus=true#step:7:69

**Comment by adamdmharvey** [comment]
I think this view shows all the FOSSA workflows, where you can see the failures:
https://github.com/backstage/backstage/actions?query=workflow%3AFOSSA

**Pull request merged**